src/fairseq2/gang.py

```diff
@@ -119,79 +119,125 @@ class ProcessGroupGang(Gang):
-            set if the process group is initialized as part of this function.
+            The timeout for operations executed against the process group.
         :param warn_only:
             If ``True``, logs a warning instead of raising an error if the
             process group is not set up reliably.
+        :param ok_initialized:
+            If ``True``, does not raise an error if the process group is already
+            initialized.
         """
         if not dist.is_available():
             raise RuntimeError("`torch.distributed` is not available.")
 
         if dist.is_initialized():
-            backend = dist.get_backend()
+            if ok_initialized:
+                return ProcessGroupGang.from_default_process_group()
 
-            if backend == "gloo":
-                device = CPU
-            elif backend == "nccl":
-                device = _determine_default_cuda_device()
-            else:
-                raise RuntimeError(
-                    f"Only `nccl` and `gloo` backends are supported for device selection, but the process group uses the `{backend}` backend."
-                )
-        else:
-            device = _determine_default_device()
+            raise RuntimeError("The default process group is already initialized.")
 
-            if timeout is None:
-                timeout = timedelta(minutes=15)
+        num_workers = _get_num_workers()
 
-            dist.init_process_group(
-                "nccl" if device.type == "cuda" else "gloo", timeout=timeout
-            )
+        if num_workers > 1 and "OMP_NUM_THREADS" not in os.environ:
+            # To prevent thread oversubscription, we distribute cores evenly
+            # across workers.
+            num_threads = max(_get_num_cores() // num_workers, 1)
 
-        if dist.group.WORLD is None:
+            torch.set_num_threads(num_threads)
+
+        if device is None:
+            device = _determine_default_device()
+
+            assert device.type == "cpu" or device.type == "cuda"
+
+        if device.type == "cpu":
+            backend = "gloo"
+        elif device.type == "cuda":
+            backend = "nccl"
+        else:
             raise RuntimeError(
-                "The default process group is not available. Please file a bug report."
+                f"Only CPU and CUDA devices are supported, but `device` is a {device.type.upper()} device."
             )
 
         if device.type == "cuda" and "NCCL_ASYNC_ERROR_HANDLING" not in os.environ:
             if warn_only:
                 logger.warning(
-                    "The default process group uses the NCCL backend, but the `NCCL_ASYNC_ERROR_HANDLING` environment variable is not set. Your collective communication calls can hang indefinitely."
+                    "The default process group will use the NCCL backend, but the `NCCL_ASYNC_ERROR_HANDLING` environment variable is not set. Your collective communication calls can hang indefinitely."
```

Comment:
Nit: maybe we should link to https://github.com/pytorch/pytorch/issues/46874, as a "learn more about it this variable at ..."